### PR TITLE
Reuse the ConfigImpl held by RdsRouteConfigProviderImpl in SRDS

### PR DESCRIPTION
**Pull request by stevenzzzz**
Signed-off-by: Xin Zhuang <stevenzzz@google.com>
See details in https://github.com/envoyproxy/envoy/issues/18240

Commit Message: Reuse the ConfigImpl from RDS config provider in SRDS.
Additional Description:
Risk Level: LOW, refactor
Testing: reuse existing tests.
Docs Changes:
Release Notes:

### CODE REVIEW

File: source/common/router/scoped_rds.cc [hunk 1]
```diff
@@ -393,16 +393,13 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
...
   auto iter = scoped_route_map_.find(scope_name);
   ASSERT(iter != scoped_route_map_.end(),
          fmt::format("trying to update route config for non-existing scope {}", scope_name));
   auto new_scoped_route_info = std::make_shared<ScopedRouteInfo>(
       envoy::config::route::v3::ScopedRouteConfiguration(iter->second->configProto()),
-      std::make_shared<ConfigImpl>(
-          rds_subscription.routeConfigUpdate()->protobufConfiguration(), optional_http_filters_,
-          factory_context_, factory_context_.messageValidationContext().dynamicValidationVisitor(),
-          false));
+      new_rds_config);
```
Comment (chaoqin-li1123) [inline source/common/router/scoped_rds.cc:402]: Can we move the new_rds_config here?

**Comment by stevenzzzz** [comment]
/assign @yanavlasov 
hi Yan, could you pls take a look?

File: source/common/router/scoped_rds.cc [hunk 2]
```diff
@@ -393,16 +393,13 @@ void ScopedRdsConfigSubscription::onConfigUpdate(
...
-                                                    RdsRouteConfigSubscription& rds_subscription) {
+                                                    ConfigConstSharedPtr new_rds_config) {
   auto iter = scoped_route_map_.find(scope_name);
   ASSERT(iter != scoped_route_map_.end(),
          fmt::format("trying to update route config for non-existing scope {}", scope_name));
   auto new_scoped_route_info = std::make_shared<ScopedRouteInfo>(
       envoy::config::route::v3::ScopedRouteConfiguration(iter->second->configProto()),
-      std::make_shared<ConfigImpl>(
-          rds_subscription.routeConfigUpdate()->protobufConfiguration(), optional_http_filters_,
-          factory_context_, factory_context_.messageValidationContext().dynamicValidationVisitor(),
```
Comment (lambdai) [inline source/common/router/scoped_rds.cc:404]: Can you confirm the Config doesn't hold a factory_context_ from another networkfilterchain/listener?
Reuse a protobuf nessage is pretty safe but Config is risk when the config is shared among different filter chain. There were a lot of issues

Comment (chaoqin-li1123) [reply source/common/router/scoped_rds.cc:404]: I think they all use the factory_context_ of [ScopedRdsConfigSubscription.](https://github.com/envoyproxy/envoy/blob/6b18c923fae9d583d80de348f9656d7420ab8bb8/source/common/router/scoped_rds.cc#L194)

Comment (lambdai) [reply source/common/router/scoped_rds.cc:404]: LGTM. Thank you for hte explanation!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**